File: src/shared/roadtrip/collage.ts
/**
 * A slide that holds SEVERAL pictures — the document side of the collage
 * frames. Pure and DOM-free; the geometry is `shared/media/media-layout.ts`,
 * the paint is `shared/media/cell-paint.ts`.
 *
 * The shape that keeps every one-picture reader working: **cell 1 IS the
 * slide.** Its `media`, `framing` and `develop` stay exactly where they are
 * today, so the rail's thumbnail, the Library sync, the date read from the
 * picture and the export plan never learn that a collage exists; the collage
 * carries cells 2…n in `cells`, each with its own picture, framing and
 * develop — never inherited, the rule every crop in this tool follows.
 *
 * A collage keeps MORE cells than its template draws: switching from a
 * six-piece bento to a 2 × 2 keeps pictures five and six in the list, not
 * drawn, and a bigger layout brings them back. Nothing is dropped by a click
 * on a tile.
 *
 * An empty cell draws nothing in the export. The editor draws a slot over it
 * on its own overlay, never in the picture.
 */

import { type Framing, normaliseFraming } from '../media/framing';
import {
  isLayoutId,
  layoutTemplate,
  type LayoutTemplateEntry,
} from '../media/layout-templates';
import {
  DEFAULT_CELL_PLACE,
  cellCount,
  normaliseCellPlace,
  normaliseSpacing,
  resolveLayout,
  type CellPlace,
  type CellRect,
  type LayoutSpacing,
} from '../media/media-layout';
import { developOrNull, type DevelopSettings } from '../develop/develop';
import type { SavedMediaRef } from '../projects/project-types';
import { transformAt, type AnimStep } from '../overlay/animation';
import { normaliseStagger, staggerDelays, staggerRanks, staggerSettle, type Stagger } from '../overlay/stagger';
import type { CellMotion } from '../media/cell-paint';
import { readAnimStep } from './badge-layout';

/** One cell after the first: a picture, how it sits, how it is corrected. */
export interface CollageCell {
  media: SavedMediaRef | null;
  framing: Framing;
  develop: DevelopSettings | null;
  /** Where the author moved this print, on a free layout; the template's place otherwise. */
  place: CellPlace;
}

/** How the cells ARRIVE: one step for all of them, spread by where they sit. */
export interface CollageEnter {
  step: AnimStep;
  stagger: Stagger;
}

/**
 * How the cells LEAVE, laid against the slide's screen time. `delay` is an
 * entrance's alone in the engine, so a staggered exit is expressed as each
 * cell's WINDOW ending earlier — the last to arrive leaves first when
 * `reverse` is set, in arrival order otherwise.
 */
export interface CollageExit {
  step: AnimStep;
  reverse: boolean;
}

export interface SlideCollage {
  /** A `layout-templates.ts` id; an id this build does not know reads as no collage. */
  template: string;
  spacing: LayoutSpacing;
  /** Painted where no picture covers the frame — between cells and in an empty one. */
  background: string;
  /** Cells 2…n. The slide's own picture is cell 1. */
  cells: CollageCell[];
  /** Where the SLIDE's own print was moved, on a free layout. */
  place: CellPlace;
  /** The cells' entrance, or null for cells that are simply there. */
  enter: CollageEnter | null;
  /** The cells' exit, or null for cells that stay to the slide's end. */
  exit: CollageExit | null;
}

export function defaultCollageEnter(): CollageEnter {
  return {
    step: { preset: 'slide', duration: 0.5, easing: 'out-cubic', direction: 'up', distanceFrac: 0.06, inside: true },
    stagger: { each: 0.1, order: 'center-out' },
  };
}

export function defaultCollageExit(): CollageExit {
  return { step: { preset: 'fade', duration: 0.4, easing: 'in' }, reverse: true };
}

/** Mirror an entrance into an exit: the same step, travelling back the way it came. */
export function mirroredExit(enter: CollageEnter): CollageExit {
  const flip = { up: 'down', down: 'up', left: 'right', right: 'left' } as const;
  const direction = enter.step.direction ? flip[enter.step.direction] : undefined;
  return {
    step: { ...enter.step, delay: undefined, ...(direction ? { direction } : {}) },
    reverse: true,
  };
}

function readEnter(v: unknown): CollageEnter | null {
  if (!v || typeof v !== 'object') return null;
  const e = v as Partial<CollageEnter>;
  if (!e.step || typeof e.step !== 'object') return null;
  const step = readAnimStep(e.step);
  if ((e.step as AnimStep).inside === true) step.inside = true;
  return { step, stagger: normaliseStagger(e.stagger) };
}

function readExit(v: unknown): CollageExit | null {
  if (!v || typeof v !== 'object') return null;
  const e = v as Partial<CollageExit>;
  if (!e.step || typeof e.step !== 'object') return null;
  const step = readAnimStep(e.step);
  if ((e.step as AnimStep).inside === true) step.inside = true;
  return { step, reverse: e.reverse !== false };
}

/** True when the cells move at all — what makes a collage slide a video. */
export function collageAnimates(collage: SlideCollage | null | undefined): boolean {
  if (!collage) return false;
  const enter = collage.enter ?? null;
  const exit = collage.exit ?? null;
  const moving = (e: { step: AnimStep; stagger?: Stagger } | null) =>
    Boolean(e && (e.step.preset !== 'none' || (e.stagger && e.stagger.each > 0)));
  return moving(enter) || Boolean(exit);
}

/**
 * Every drawn cell's motion at `t` seconds into the slide, or null for a
 * collage that does not move. `seconds` is the slide's screen time — what an
 * exit is laid against; without one, cells that entered stay.
 */
export function collageCellMotions(
  collage: SlideCollage,
  cells: readonly CellRect[],
  frame: { w: number; h: number },
  t: number,
  seconds: number | null,
): (CellMotion | null)[] | null {
  const enter = collage.enter ?? null;
  const exit = collage.exit ?? null;
  if (!enter && !exit) return null;
  const delays = enter ? staggerDelays(cells, frame, enter.stagger) : cells.map(() => 0);
  let ends: (number | null)[] = cells.map(() => null);
  if (exit && seconds !== null) {
    const stagger = enter?.stagger ?? { each: 0, order: 'sequence' as const };
    const ranks = staggerRanks(cells, frame, stagger.order, stagger.seed ?? 0);
    const top = Math.max(0, ...ranks);
    // Reverse: last in, first out — the highest rank's window ends first.
    ends = ranks.map((r) => seconds - (exit.reverse ? r : top - r) * Math.max(0, stagger.each));
  }
  return cells.map((_, i) => {
    const anim = {
      in: enter ? { ...enter.step, delay: delays[i] } : null,
      out: exit && ends[i] !== null ? exit.step : null,
    };
    const transform = transformAt(anim, { start: 0, end: ends[i] }, t);
    const step = t < (delays[i] + (enter?.step.duration ?? 0)) ? enter?.step : exit?.step;
    return { transform, direction: step?.direction, inside: step?.inside };
  });
}

/**
 * When a collage slide is at rest — the last cell's entrance done. What a
 * still of it is taken at; 0 with no entrance. Resolved on a frame of the
 * slide's shape, since the ranks depend on where the cells are.
 */
export function collageSettleSeconds(collage: SlideCollage | null | undefined, aspect: number): number {
  const enter = collage?.enter ?? null;
  if (!collage || !enter) return 0;
  const cells = resolveCollage(collage, aspect, 1);
  return staggerSettle(staggerDelays(cells, { w: aspect, h: 1 }, enter.stagger), enter.step);
}

/** The suite's frame black — what a stage shows behind a picture that does not cover it. */
export const DEFAULT_COLLAGE_BACKGROUND = '#100f0d';

export function createCollageCell(media: SavedMediaRef | null = null): CollageCell {
  return {
    media,
    framing: normaliseFraming(undefined),
    develop: null,
    place: { ...DEFAULT_CELL_PLACE },
  };
}

/** A new collage on `template`, its cells 2…n empty. Null for an unknown id. */
export function createCollage(template: string): SlideCollage | null {
  const entry = layoutTemplate(template);
  if (!entry) return null;
  const count = cellCount(entry.template);
  return {
    template,
    spacing: normaliseSpacing(undefined),
    background: DEFAULT_COLLAGE_BACKGROUND,
    cells: Array.from({ length: Math.max(0, count - 1) }, () => createCollageCell()),
    place: { ...DEFAULT_CELL_PLACE },
    enter: null,
    exit: null,
  };
}

function mediaRefOrNull(v: unknown): SavedMediaRef | null {
  if (!v || typeof v !== 'object') return null;
  const r = v as Partial<SavedMediaRef>;
  if (typeof r.name !== 'string' || !r.name) return null;
  const ref: SavedMediaRef = {
    name: r.name,
    size: typeof r.size === 'number' && Number.isFinite(r.size) ? r.size : 0,
    lastModified:
      typeof r.lastModified === 'number' && Number.isFinite(r.lastModified) ? r.lastModified : 0,
  };
  if (typeof r.assetId === 'string' && r.assetId) ref.assetId = r.assetId;
  if (typeof r.hash === 'string' && r.hash) ref.hash = r.hash;
  return ref;
}

function colour(v: unknown): string {
  return typeof v === 'string' && /^#[0-9a-f]{6}$/i.test(v.trim())
    ? v.trim().toLowerCase()
    : DEFAULT_COLLAGE_BACKGROUND;
}

/**
 * Read a collage out of anything — a stored document, an imported file, a
 * hand edit. A template the registry does not know is NO collage: a slide
 * written by a newer Atelier opens here as its lead picture and loses the
 * layout, it never fails to open. Junk in a cell lands clamped or empty.
 */
export function readCollage(v: unknown): SlideCollage | null {
  if (!v || typeof v !== 'object') return null;
  const c = v as Partial<SlideCollage>;
  if (!isLayoutId(c.template)) return null;
  const cells = Array.isArray(c.cells) ? c.cells : [];
  return {
    template: c.template,
    spacing: normaliseSpacing(c.spacing),
    background: colour(c.background),
    cells: cells.map((cell) => {
      const raw = (cell && typeof cell === 'object' ? cell : {}) as Partial<CollageCell>;
      return {
        media: mediaRefOrNull(raw.media),
        framing: normaliseFraming(raw.framing),
        develop: developOrNull(raw.develop),
        place: normaliseCellPlace(raw.place),
      };
    }),
    place: normaliseCellPlace(c.place),
    enter: readEnter(c.enter),
    exit: readExit(c.exit),
  };
}

/** The registry entry a collage draws with. Null only for an id `readCollage` would already have refused. */
export function collageEntry(collage: SlideCollage): LayoutTemplateEntry | null {
  return layoutTemplate(collage.template);
}

/** How many cells the collage's template DRAWS (its `cells` may hold more). */
export function collageCellCount(collage: SlideCollage): number {
  const entry = collageEntry(collage);
  return entry ? cellCount(entry.template) : 0;
}

/**
 * The lead picture as the collage's first cell — what lets every consumer
 * treat "cell i" uniformly. Cell 0 is the slide; the rest are `cells[i − 1]`.
 */
export interface CollageLead {
  media: SavedMediaRef | null;
  framing: Framing;
  develop: DevelopSettings | null;
}

export function collageCellAt(lead: CollageLead, collage: SlideCollage, i: number): CollageCell {
  if (i === 0) return { media: lead.media, framing: lead.framing, develop: lead.develop, place: collage.place };
  return collage.cells[i - 1] ?? createCollageCell();
}

/** The places of every drawn cell, in cell order, for `resolveLayout`. */
export function collagePlaces(collage: SlideCollage): CellPlace[] {
  const n = collageCellCount(collage);
  return Array.from({ length: n }, (_, i) => (i === 0 ? collage.place : (collage.cells[i - 1]?.place ?? DEFAULT_CELL_PLACE)));
}

/** The collage's cells in a `w`×`h` frame. Empty for an unknown template. */
export function resolveCollage(collage: SlideCollage, w: number, h: number): CellRect[] {
  const entry = collageEntry(collage);
  if (!entry) return [];
  return resolveLayout(entry.template, w, h, collage.spacing, collagePlaces(collage));
}

/**
 * The refs every DRAWN cell names, lead first — what a surface needs to find
 * or fetch before it can paint the slide. A cell without a picture is skipped.
 */
export function collageMediaRefs(lead: CollageLead, collage: SlideCollage): SavedMediaRef[] {
  const n = collageCellCount(collage);
  const refs: SavedMediaRef[] = [];
  for (let i = 0; i < n; i++) {
    const ref = collageCellAt(lead, collage, i).media;
    if (ref) refs.push(ref);
  }
  return refs;
}

/**
 * The pictures a smaller template no longer draws — kept, and listed so the
 * author knows they are there. Indices are 1-based cell numbers.
 */
export function collageKept(collage: SlideCollage): { cell: number; media: SavedMediaRef }[] {
  const n = collageCellCount(collage);
  const kept: { cell: number; media: SavedMediaRef }[] = [];
  collage.cells.forEach((cell, k) => {
    const number = k + 2;
    if (number > n && cell.media) kept.push({ cell: number, media: cell.media });
  });
  return kept;
}

/**
 * Change the template and keep every picture: the list is padded with empty
 * cells up to the new count, never truncated.
 */
export function retemplateCollage(collage: SlideCollage, template: string): SlideCollage | null {
  const entry = layoutTemplate(template);
  if (!entry) return null;
  const need = Math.max(0, cellCount(entry.template) - 1);
  const cells = collage.cells.slice();
  while (cells.length < need) cells.push(createCollageCell());
  return { ...collage, template, cells };
}

/** Swap two cells' pictures, framings and develops — never their places, which belong to the slot. */
export function swapCollageCells(
  lead: CollageLead,
  collage: SlideCollage,
  a: number,
  b: number,
): { lead: CollageLead; collage: SlideCollage } {
  if (a === b || a < 0 || b < 0) return { lead, collage };
  const n = Math.max(collage.cells.length + 1, a + 1, b + 1);
  const all: CollageCell[] = Array.from({ length: n }, (_, i) => collageCellAt(lead, collage, i));
  const [A, B] = [all[a], all[b]];
  const swap = (from: CollageCell, to: CollageCell): CollageCell => ({
    ...to,
    media: from.media,
    framing: from.framing,
    develop: from.develop,
  });
  all[a] = swap(B, A);
  all[b] = swap(A, B);
  return {
    lead: { media: all[0].media, framing: all[0].framing, develop: all[0].develop },
    collage: { ...collage, cells: all.slice(1) },
  };
}

/** Write cell `i` (0 = the lead), returning the new lead and collage. */
export function withCollageCell(
  lead: CollageLead,
  collage: SlideCollage,
  i: number,
  patch: Partial<CollageCell>,
): { lead: CollageLead; collage: SlideCollage } {
  if (i === 0) {
    return {
      lead: {
        media: patch.media !== undefined ? patch.media : lead.media,
        framing: patch.framing ?? lead.framing,
        develop: patch.develop !== undefined ? patch.develop : lead.develop,
      },
      collage: patch.place ? { ...collage, place: patch.place } : collage,
    };
  }
  const cells = collage.cells.slice();
  while (cells.length < i) cells.push(createCollageCell());
  cells[i - 1] = { ...cells[i - 1], ...patch };
  return { lead, collage: { ...collage, cells } };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Roadtrip/Collage.swift
// A slide that holds SEVERAL pictures, as the trip document stores it — port
// of the stored half of `src/shared/roadtrip/collage.ts`: the collage record,
// its cells, how they arrive and leave, the factories and the reader.
//
// Types + reader only; the behaviour of `collage.ts` (which cells a template
// draws, the cells' rects, their motion over the slide, the settle time, the
// kept pictures, re-templating, swapping and writing a cell) is ported later
// INTO THIS FILE. The geometry is `Media/MediaLayout.swift`'s.
//
// Rules kept:
// - **Cell 1 IS the slide.** Its picture, framing, develop and motion stay on
//   the slide itself, so every one-picture reader keeps working; `cells`
//   holds cells 2…n, each with its own picture, framing and develop — never
//   inherited.
// - A collage keeps MORE cells than its template draws: a smaller layout
//   keeps the extra pictures in the list, not drawn.
// - A template this build does not know is NO collage: a slide written by a
//   newer Atelier opens as its lead picture, it never fails to open. Junk in
//   a cell lands clamped or empty.

import Foundation

/// One cell after the first: a picture, how it sits, how it is corrected.
public struct CollageCell: Equatable, Sendable {
    public var media: SavedMediaRef?
    public var framing: Framing
    public var develop: DevelopSettings?
    /// How this cell's picture MOVES inside its mask over the slide, or nil.
    public var motion: FramingMotion?
    /// Where the author moved this print, on a free layout; the template's place otherwise.
    public var place: CellPlace

    public init(media: SavedMediaRef? = nil, framing: Framing = .default, develop: DevelopSettings? = nil,
                motion: FramingMotion? = nil, place: CellPlace = defaultCellPlace) {
        self.media = media; self.framing = framing; self.develop = develop; self.motion = motion; self.place = place
    }

    public var json: JSONValue {
        .object([
            "media": media?.json ?? .null, "framing": framing.json, "develop": develop?.json ?? .null,
            "motion": motion?.json ?? .null, "place": place.json,
        ])
    }
}

/// How the cells ARRIVE: one step for all of them, spread by where they sit.
public struct CollageEnter: Equatable, Sendable {
    public var step: AnimStep
    public var stagger: Stagger

    public init(step: AnimStep, stagger: Stagger) {
        self.step = step
        self.stagger = stagger
    }

    public var json: JSONValue {
        .object(["step": step.json, "stagger": stagger.json])
    }
}

/// How the cells LEAVE, laid against the slide's screen time — the last to
/// arrive leaves first when `reverse` is set.
public struct CollageExit: Equatable, Sendable {
    public var step: AnimStep
    public var reverse: Bool

    public init(step: AnimStep, reverse: Bool) {
        self.step = step
        self.reverse = reverse
    }

    public var json: JSONValue {
        .object(["step": step.json, "reverse": .bool(reverse)])
    }
}

public struct SlideCollage: Equatable, Sendable {
    /// A `LayoutTemplates.swift` id; one this build does not know reads as no collage.
    public var template: String
    public var spacing: LayoutSpacing
    /// Painted where no picture covers the frame — between cells and in an empty one.
    public var background: String
    /// Cells 2…n. The slide's own picture is cell 1.
    public var cells: [CollageCell]
    /// Where the SLIDE's own print was moved, on a free layout.
    public var place: CellPlace
    /// The cells' entrance, or nil for cells that are simply there.
    public var enter: CollageEnter?
    /// The cells' exit, or nil for cells that stay to the slide's end.
    public var exit: CollageExit?

    public init(template: String, spacing: LayoutSpacing = defaultLayoutSpacing, background: String = defaultCollageBackground,
                cells: [CollageCell] = [], place: CellPlace = defaultCellPlace, enter: CollageEnter? = nil, exit: CollageExit? = nil) {
        self.template = template; self.spacing = spacing; self.background = background; self.cells = cells
        self.place = place; self.enter = enter; self.exit = exit
    }

    /// The collage as the document holds it: nulls for no entrance and no exit.
    public var json: JSONValue {
        .object([
            "template": .string(template), "spacing": spacing.json, "background": .string(background),
            "cells": .array(cells.map(\.json)), "place": place.json,
            "enter": enter?.json ?? .null, "exit": exit?.json ?? .null,
        ])
    }
}

/// The suite's frame black — what a stage shows behind a picture that does not cover it.
public let defaultCollageBackground = "#100f0d"

public func defaultCollageEnter() -> CollageEnter {
    CollageEnter(
        step: AnimStep(preset: .slide, duration: 0.5, easing: .outCubic, direction: .up, distanceFrac: 0.06, inside: true),
        stagger: Stagger(each: 0.1, order: .centerOut)
    )
}

public func defaultCollageExit() -> CollageExit {
    CollageExit(step: AnimStep(preset: .fade, duration: 0.4, easing: .in), reverse: true)
}

public func createCollageCell(_ media: SavedMediaRef? = nil) -> CollageCell {
    CollageCell(media: media, framing: normaliseFraming(nil), develop: nil, motion: nil, place: defaultCellPlace)
}

/// A new collage on `template`, its cells 2…n empty. Nil for an unknown id.
public func createCollage(_ template: String) -> SlideCollage? {
    guard let entry = layoutTemplate(template) else { return nil }
    let count = cellCount(entry.template)
    return SlideCollage(
        template: template,
        spacing: normaliseSpacing(nil),
        background: defaultCollageBackground,
        cells: (0..<max(0, count - 1)).map { _ in createCollageCell() },
        place: defaultCellPlace,
        enter: nil,
        exit: nil
    )
}

/// A ref read the collage's way: a name is required; a size or a date that is
/// not a finite number is 0; an empty id or hash is left out.
private func collageMediaRef(_ raw: JSONValue?) -> SavedMediaRef? {
    guard let r = raw?.objectValue, let name = r["name"]?.stringValue, !name.isEmpty else { return nil }
    let size = r["size"]?.finiteNumber ?? 0
    var ref = SavedMediaRef(name: name, size: abs(size) < 9e15 ? Int(size) : 0,
                            lastModified: r["lastModified"]?.finiteNumber ?? 0)
    if let assetId = r["assetId"]?.stringValue, !assetId.isEmpty { ref.assetId = assetId }
    if let hash = r["hash"]?.stringValue, !hash.isEmpty { ref.hash = hash }
    return ref
}

/// `/^#[0-9a-f]{6}$/i` after a trim, kept lower-cased; anything else is the frame black.
private func collageColour(_ raw: JSONValue?) -> String {
    guard let s = raw?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines) else { return defaultCollageBackground }
    let bytes = Array(s.utf8)
    guard bytes.count == 7, bytes[0] == UInt8(ascii: "#") else { return defaultCollageBackground }
    for b in bytes[1...] {
        let digit = (b >= 48 && b <= 57) || (b >= 65 && b <= 70) || (b >= 97 && b <= 102)
        if !digit { return defaultCollageBackground }
    }
    return s.lowercased()
}

/// A step and what rides with it: the badge's reader, plus `inside` when it is exactly true.
private func collageStep(_ raw: JSONValue) -> AnimStep {
    var step = readAnimStep(raw)
    if raw.objectValue?["inside"] == .bool(true) { step.inside = true }
    return step
}

/// A record holding a step (a record — or a list, which the web's `typeof`
/// takes for one), or nil.
private func stepOf(_ raw: JSONValue?) -> (record: [String: JSONValue], step: JSONValue)? {
    guard let e = raw?.objectValue, let step = e["step"], step.objectValue != nil || step.arrayValue != nil else { return nil }
    return (e, step)
}

private func readEnter(_ raw: JSONValue?) -> CollageEnter? {
    guard let found = stepOf(raw) else { return nil }
    return CollageEnter(step: collageStep(found.step), stagger: normaliseStagger(found.record["stagger"]))
}

private func readExit(_ raw: JSONValue?) -> CollageExit? {
    guard let found = stepOf(raw) else { return nil }
    return CollageExit(step: collageStep(found.step), reverse: found.record["reverse"] != .bool(false))
}

/// Read a collage out of anything — a stored document, an imported file, a
/// hand edit. A template the registry does not know is NO collage.
public func readCollage(_ v: JSONValue?) -> SlideCollage? {
    guard let c = v?.objectValue, let template = c["template"]?.stringValue, layoutTemplate(template) != nil else { return nil }
    let cells = (c["cells"]?.arrayValue ?? []).map { cell -> CollageCell in
        let raw = cell.objectValue ?? [:]
        return CollageCell(
            media: collageMediaRef(raw["media"]),
            framing: normaliseFraming(raw["framing"]),
            develop: developOrNull(raw["develop"]),
            motion: readMotion(raw["motion"]),
            place: normaliseCellPlace(raw["place"])
        )
    }
    return SlideCollage(
        template: template,
        spacing: normaliseSpacing(c["spacing"]),
        background: collageColour(c["background"]),
        cells: cells,
        place: normaliseCellPlace(c["place"]),
        enter: readEnter(c["enter"]),
        exit: readExit(c["exit"])
    )
}
